modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -76,47 +65,57 @@
  * their tests can use them. Most methods run in the listener thread pool because the are meant to be fast and don't expect to block.
  */
 public abstract class AbstractAsyncBulkByScrollAction<Request extends AbstractBulkByScrollRequest<Request>> {
+    protected final ESLogger logger;
+    protected final BulkByScrollTask task;
+    protected final ThreadPool threadPool;
     /**
      * The request for this action. Named mainRequest because we create lots of <code>request</code> variables all representing child
      * requests of this mainRequest.
      */
     protected final Request mainRequest;
-    protected final BulkByScrollTask task;
 
     private final AtomicLong startTime = new AtomicLong(-1);
-    private final AtomicReference<String> scroll = new AtomicReference<>();
     private final AtomicLong lastBatchStartTime = new AtomicLong(-1);
     private final Set<String> destinationIndices = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
-    private final ESLogger logger;
     private final ParentTaskAssigningClient client;
-    private final ThreadPool threadPool;
-    private final SearchRequest firstSearchRequest;
     private final ActionListener<BulkIndexByScrollResponse> listener;
-    private final BackoffPolicy backoffPolicy;
     private final Retry bulkRetry;
+    private final ScrollableHitSource scrollSource; 
 
     public AbstractAsyncBulkByScrollAction(BulkByScrollTask task, ESLogger logger, ParentTaskAssigningClient client,
-                                           ThreadPool threadPool, Request mainRequest, SearchRequest firstSearchRequest,
-                                           ActionListener<BulkIndexByScrollResponse> listener) {
+                                           ThreadPool threadPool, Request mainRequest, ActionListener<BulkIndexByScrollResponse> listener) {
         this.task = task;
         this.logger = logger;
         this.client = client;
         this.threadPool = threadPool;
         this.mainRequest = mainRequest;
-        this.firstSearchRequest = firstSearchRequest;
         this.listener = listener;
-        backoffPolicy = buildBackoffPolicy();
+        BackoffPolicy backoffPolicy = buildBackoffPolicy();
         bulkRetry = Retry.on(EsRejectedExecutionException.class).policy(wrapBackoffPolicy(backoffPolicy));
```

Comment:
Do you think we could have something like:
```java
bulkRetry = Retry.on(EsRejectedExecutionException.class).policy(BackoffPolicy.wrap(backoffPolicy, task::countBulkRetry));
```

I find it easier to know what's going on on bulk retries.